02_knight_and_princess/anton_potapov/graph_create.cpp

```diff
@@ -0,0 +1,175 @@
+#include <cassert>
+#include <fstream>
+#include <iostream>
+#include <map>
+#include <set>
+#include <sstream>
+#include <stdexcept>
+#include <string>
+#include <vector>
+
+using VertexId = int;
+using EdgeId = int;
+
+class Vertex {
+ public:
+  explicit Vertex(const VertexId& vertex_id) : id_(vertex_id) {}
+
+  const std::set<EdgeId>& connected_edges() const { return connected_edges_; }
+
+  std::string get_json_string() const {
+    std::stringstream json_stringstream;
+    json_stringstream << "{\"id\":" << id_ << ","
+                      << "\"edge_ids\":[";
+    for (auto it = connected_edges_.begin(); it != connected_edges_.end();
+         ++it) {
+      json_stringstream << *it;
+      if (std::next(it) != connected_edges_.end()) {
+        json_stringstream << ",";
+      }
+    }
+    json_stringstream << "]}";
+    return json_stringstream.str();
+  }
+  bool has_edge_id(const EdgeId& edge_id) const {
+    return connected_edges_.find(edge_id) != connected_edges_.end();
+  }
+  void add_edge(const EdgeId& edge_id) {
+    assert(!has_edge_id(edge_id) && "edge that is to be added already exists");
+    connected_edges_.insert(edge_id);
+  }
+
+ private:
+  const VertexId id_;
+  std::set<EdgeId> connected_edges_;
+};
+
+class Edge {
+ public:
+  const VertexId vertex1_id, vertex2_id;
+  Edge(const EdgeId& edge_id, const VertexId& vertex1, const VertexId& vertex2)
+      : vertex1_id(vertex1), vertex2_id(vertex2), id_(edge_id) {}
+  std::string get_json_string() const {
+    std::stringstream json_stringstream;
+    json_stringstream << "{\"id\":" << id_ << ","
+                      << "\"vertex_ids\":[" << vertex1_id << "," << vertex2_id
+                      << "]}";
+    return json_stringstream.str();
+  }
+
+ private:
+  const EdgeId id_;
+};
+
+class Graph {
+ public:
+  bool is_vertex_exists(const VertexId& vertex) const {
+    return vertices_.find(vertex) != vertices_.end();
+  }
+
+  bool is_connected(const VertexId& vertex1, const VertexId& vertex2) const {
+    assert(is_vertex_exists(vertex1) && "Vertex 1 doesn't exist");
+    assert(is_vertex_exists(vertex2) && "Vertex 2 doesn't exist");
+    auto it_vertex1 = vertices_.find(vertex1);
```

Comment:
Вот видишь, уже начались проблемы из-за неправильного наименования.
Что за `it_vertex1`? Что значит `it`? Ты это написал, просто потому, что у тебя имя `vertex` уже занято. А по-хорошему оно не должно быть занято, просто ты все напутал с наименованием:
```suggestion
  bool is_connected(const VertexId& vertex1_id, const VertexId& vertex2_id) const {
    assert(is_vertex_exists(vertex1_id) && "Vertex 1 doesn't exist");
    assert(is_vertex_exists(vertex2_id) && "Vertex 2 doesn't exist");
    auto vertex1 = vertex_ids_.find(vertex1_id);
```
Видишь, насколько проще была бы читаемость кода, всего-то из-за более правильного наименования переменных.